src/agent0/utils/block_before_timestamp.py

```diff
@@ -0,0 +1,74 @@
+"""Get the number of the block that is at or immediately before the given timestamp."""
+
+from eth_typing import BlockNumber
+from web3 import Web3
+from web3.types import Timestamp
+
+# pylint: disable=too-many-locals
+
+
+def block_number_before_timestamp(web3: Web3, block_timestamp: Timestamp | int) -> BlockNumber:
+    """Finds the closest block number that is before or at the given block time.
+
+    Arguments
+    ---------
+    web3: Web3
+        The web3 instance.
+    block_timestamp: BlockTime | int
+        The block time to find the closest block to.
+
+    Returns
+    -------
+    BlockNumber
+        The closest block number to the given block time.
+    """
+    # Get the current block number and timestamp
+    block_timestamp = Timestamp(block_timestamp)
+    current_block = web3.eth.get_block("latest")
+    current_block_number = current_block.get("number", None)
+    assert current_block_number is not None
+    if current_block_number < 3:
+        raise ValueError("The current block number must be >= 2.")
+    current_block_timestamp = current_block.get("timestamp", None)
+    assert current_block_timestamp is not None
+
+    # Estimate the average block time
+    earlier_block_number = current_block_number // 2
+    earlier_block_delta = current_block_number - earlier_block_number
+    if earlier_block_delta <= 0:
+        raise ValueError("Error estimating the delta blocks.")
+    earlier_block_timestamp = web3.eth.get_block(earlier_block_number).get("timestamp", None)
+    assert earlier_block_timestamp is not None
+    avg_time_between_blocks = (current_block_timestamp - earlier_block_timestamp) / earlier_block_delta
+
+    # Estimate the block number corresponding to the user provided timestamp
+    delta_time = current_block_timestamp - block_timestamp
+    estimated_block_number = current_block_number - (delta_time // avg_time_between_blocks)
+
+    # Use a binary search to find the block
+    left = int(max(0, estimated_block_number - 100))  # search 100 blocks before estimated block
```

Comment:
Can you handle the case where we make a bad guess, and the query block timestamp is outside of initial left/right?

Reply:
```
    # Ensure bounds of binary search is within the target timestamp
    if web3.eth.get_block(left)["timestamp"] > block_timestamp:
        left = 0
    if web3.eth.get_block(right)["timestamp"] < block_timestamp:
        right = current_block_number
```